File: 2017/Haskell/02.hs
{-# LANGUAGE TypeApplications #-}
module Main where

import Data.Digits ( digits )
import Data.Semigroup
import Data.List

pairs :: [Int] -> [(Int, Int)]
pairs xs = do
  (x:xs) <- tails xs
  y <- xs
  pure (x, y)

part1 :: [[Int]] -> Int
part1 = getSum . foldMap (Sum . diff . foldMap tup)
  where
    tup x = (Max x, Min x)
    diff (Max x, Min y) = abs (x - y)

part2 :: [[Int]] -> Int
part2 = getSum . foldMap (foldMap tup . pairs)
  where
    tup (x, y) = let (a, b) = (max x y, min x y)
                  in if rem a b == 0
                        then Sum (a `div` b)
                        else Sum 0

main :: IO ()
main = do
  input <- map ((map (read @Int)) . words) . lines <$> readFile "../data/02.in"
  print $ part1 input
  print $ part2 input

File: 2017/Python/02.py

def read_file(filename):
    with open(filename) as f:
        return [[int(i) for i in line.split()] for line in f]

def checksum(lines):
    return sum(max(i) - min(i) for i in lines)

def evenly_divisible(lines):
    total = 0
    for line in lines:
        for a in line:
             for b in line:
                 if a == b:
                     continue
                 if a % b == 0:
                     total += (a // b)
    return total

def main():
    lines = read_file('input/02.in')
    print('Part 1:', checksum(lines))
    print('Part 2:', evenly_divisible(lines))

if __name__ == "__main__":
    main()

# 541 (Too high)
